### PR TITLE
#Private events -Rails Active Records Association project

### PR DESCRIPTION
**Pull request by codecell**
**What does this PR do?**
This is for TSEs to review our implementation of the private-events rails active records association project.

**Endpoints covered in this project**
**users & sessions**
- POST localhost:3000 /users/new  - For Signing up a new client
- POST localhost:3000 /users/user_id - For viewing a user's dashboard/profile
- GET localhost:3000 /signin - For logging in a user
- POST localhost:3000 /logout - For signing out a user

**events**
- POST localhost:3000 /events/new - For creating a new event
- GET localhost:3000 /events  - For viewing all events

**invitations**
- POST localhost:3000 /invitations/new - For inviting users to an event
- GET  localhost:3000 /invitations - For viewing both past and upcoming invitations.

**How can this be manually tested?**
- after cloning this repo
- run `bundle install`
- run `rails db:migrate` to get the tables setup
- start the server with the command `rails server`
- head over to the  endpoint  localhost:3000 /users/new to create a new user
- Follow the other navigation links listed on each page and/or the navbar to test out other endpoints

### CODE REVIEW

File: app/views/events/new.html.erb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<h1>New Event</h1>
+
+<%= form_for @event, :url => events_path do |f| %>
+  Title: <%= f.text_field :title %><br/>
+  Event Date: <%= f.text_field :event_date %><br/>
```
Comment (suretrust) [inline app/views/events/new.html.erb]: This should be `date_field`. Not everybody knows the format you are expecting for the date, is it `DD/MM/YYYY`? You can add something like that as a placeholder or just use `date_field` instead.

File: app/views/invitations/new.html.erb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+<h1>Invite A User To An Event</h1>
+<p><<<<<<<<<<<<<<<>>>>>>>>>>>>>>></p>
+
+ <%= form_for :invitation, url:invitations_path do |f|%>
+
+  <%= f.label :invitee_id, 'Invitee ID'%>
+  <%= f.text_field :invitee_id, class:'form-control' %> <br/>
+
+  <%= f.label :attended_event_id, 'Event ID'%>
+  <%= f.text_field :attended_event_id, class:'form-control' %> <br/>
```
Comment (suretrust) [inline app/views/invitations/new.html.erb]: How do I know the event id and user id of who I am inviting?

Screenshot 2020-06-05 at 11 08 23

File: app/views/sessions/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<%= flash[:message] %>
+
+<h1>Sing In</h1>
+
+ <%= form_for :session, url:signin_path do |f|%>
+  <%= f.label :private_event_user_id, 'User ID'%>
+  <%= f.text_field :private_event_user_id, class:'form-control' %>
+
+  <%= f.submit "Submit" %>
+<% end %>
```
Comment (suretrust) [inline app/views/sessions/new.html.erb:10]: How do  I know the ID I'm supposed to log in with? I tried with email and the name I registered with, didn't work.

Screenshot 2020-06-05 at 11 11 52

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+class User < ApplicationRecord
+  has_many :events
```
Comment (suretrust) [inline app/models/user.rb]: * [ ] User model: `has_many :events, foreign_key: :creator_id` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

File: app/models/event.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+class Event < ApplicationRecord
+  belongs_to :user
```
Comment (suretrust) [inline app/models/event.rb]: * [ ] Event model `belongs_to :creator, class_name: "User"` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
  *

**Review by suretrust — CHANGES_REQUESTED** [review]
Hi team, 👋 

### Great job thus far! 🙌 Please see my comments. 👇 

* [x] It has functional `login` page (authentication WITHOUT password, just providing a user email) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/login_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [x] The event's show page displays the list of attendants ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/events_attendees_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)). I couldn't test this because the params for inviting a user are unclear to me. More details below.
* [x] The user's show page displays events the user has created and events that the user is attending to separated in past and upcoming ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)). I was unable to confirm this for the same reason.
- [x] You should be able to track the creator of an event.

**Kindly submit for another code review request once these issues have been fixed.** _Happy coding!_ 🎉

**Review by euqueme — CHANGES_REQUESTED** [review]
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @codecell and @dansantander :balloon: :lollipop: 

### Thank you for doing most of the changes requested by the previous reviewer :medal_sports: :medal_sports: your app is working better than before and it's more intuitive to use, there are still some changes regarding the behavior and foreign key use that were required for this project so let's work on them :construction: :construction_worker_man: :construction_worker_man: 

#### Behaviour

##### Invitations
- [ ] Your app is letting a user invite users to events he/she didn't create, users should only be able to invite users to events they created.(if not then the events are not that private don't you think?)

- [ ] Join model (EventAttendance) is required to make sure that attendance is not created without invitation (e.g. `before_create` callback usage) this means when an invitation is sent the invited user should confirm the invitation before being considered an attendee of the event, your application is confirming the users just by sending the invitation.

- Invitations were optional so you can either do the above changes or remove Invitations from your project if you decide to remove them then you would have to let users attend to any event they want just by clicking a button.

#### Associations

##### Migrations
* [x] events table has creator_id as the foreign key ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

- I can see you use the user_id in your Event table but you have to create the foreign key just as it was requested by the previous reviewer because this was a requirement of the project and the objective is that you learn to work with aliases and foreign keys in rails so let's work on this change.

## [OPTIONAL]
I'm setting the following changes as optional because the previous reviewer didn't mention them but these are hard requirements of the project so please consider working on them.

#### Databases an Indexes
* [x] the events table has column `creator_id` of type `integer`
* [x] an index for `creator_id` column is added
* [x] the events_attendation/events_users/(any other name for join table between users and events) table has column `event_id` of type `integer`
* [x] an index for `event_id` column is added
* [x] the events_attendation/events_users/(any other name for join table between users and events) table has column `user_id` of type `integer`
* [x] an index for `user_id` column is added

Why indices for foreign keys are necessary? In short: Since the foreign key columns will be used often to retrieve related information, it's important to index them (queries on those columns will be faster with indexes). More details:

* https://semaphoreci.com/blog/2017/05/09/faster-rails-is-your-database-properly-indexed.html
* https://sqlperformance.com/2012/11/t-sql-queries/benefits-indexing-foreign-keys.

#### Tests

* [ ] models have unit tests for associations (Rspec) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/unit_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))

* [ ] event managment features have integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))

- Also, create an Automated Test section and add how to run your test cases in your Readme file 

### That's it, team! :raised_hands: 

### Please :pray: do the changes :technologist: :technologist: and submit back for a code review when you are done :memo: 

Happy coding! :rocket: 

Any questions? ask me on slack **@ Maru** :woman_technologist: 

PROJECT STATUS: REQUESTED CHANGES :vertical_traffic_light:

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi Team       
#### Great Job so far, but there are a few more things required. 
### General
- [x] It has functional `login` page ( just providing a user email not ID)
- [x]  Models have unit tests for associations (Rspec)
- [x] Authentication have has integration tests (Rspec + Capybara)
- [x]  Event managment features have integration tests
### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Comment by codecell** [comment]
Hi @kobiyoyo , 
We can't easily change the sign-in functionality to use **email** because the project specifically said we can use **id** or even a ** user's name**  see below;
image
[Odin Link](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)
And again, it will require us to change so many things which considering the amount of work we've put in would not be easy.

**Review by suretrust — APPROVED** [review]
Hi team, 👋 

### Great job implementing the suggested changes. No deal-breaker, let's get this merged. 👍 

**Note:** In a real-life situation, you can't expect a user to login with an ID that he didn't register with. For best UX, user should log in with parameters that they signed up with. In the instruction on Odin, the ID is synonymous with username, not the ID of the user from the user's table. Users will definitely forget the ID they saw at a glace after signing up because I'm sure they didn't even pay much attention to it.

_Good luck with your future projects and happy coding!_ 🎉

**Pull request merged**